### DISCUSSION TITLE
Is next-auth compatible with NextJS Edge Runtime ?

### DISCUSSION THREAD

**Discussion by rupampoddar** [nextauthjs/next-auth#5855] [Help] [answered]
Build is failing with the following error

```bash
$ pnpm run build

> my-app@ build /mnt/data/projects/my-app
> next build

info  - Loaded env from /mnt/data/projects/my-app/.env
warn  - You have enabled experimental feature (runtime) in next.config.js.
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.

info  - Linting and checking validity of types  
warn  - You are using the experimental Edge Runtime with `experimental.runtime`.
info  - Creating an optimized production build  
Failed to compile.

./node_modules/.pnpm/@babel+runtime@7.20.1/node_modules/@babel/runtime/regenerator/index.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime 
Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation

Import trace for requested module:
./node_modules/.pnpm/@babel+runtime@7.20.1/node_modules/@babel/runtime/regenerator/index.js
./node_modules/.pnpm/next-auth@4.17.0_dcors23iqtgxrhrwosgnnc4xji/node_modules/next-auth/core/errors.js
./node_modules/.pnpm/next-auth@4.17.0_dcors23iqtgxrhrwosgnnc4xji/node_modules/next-auth/utils/logger.js
./node_modules/.pnpm/next-auth@4.17.0_dcors23iqtgxrhrwosgnnc4xji/node_modules/next-auth/core/index.js
./node_modules/.pnpm/next-auth@4.17.0_dcors23iqtgxrhrwosgnnc4xji/node_modules/next-auth/next/index.js
./node_modules/.pnpm/next-auth@4.17.0_dcors23iqtgxrhrwosgnnc4xji/node_modules/next-auth/index.js


> Build failed because of webpack errors
 ELIFECYCLE  Command failed with exit code 1.
```

`next.config.js`

```javascript
/**
 * @type {import('next').NextConfig}
 */
module.exports = {
  reactStrictMode: true,
  swcMinify: true,
  experimental:{
    runtime:"experimental-edge"
  },
};

```

`pages/api/auth/[...nextauth].ts`

```javascript
import NextAuth, { NextAuthOptions } from 'next-auth'
import GoogleProvider from 'next-auth/providers/google'

export const authOptions: NextAuthOptions = {
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
  ],
  theme: {
    colorScheme: 'light',
  },
}

export default NextAuth(authOptions)

```

Is this a known limitation or am I doing something wrong ?

**Comment by az-hussain** [comment]
Commenting to follow along, was hoping to implement as a solution to the (sometimes long) delay between Signup() and actually redirecting user to the /login page. 

Vercel's Edge documentation points out, however, that "you cannot use Node.js APIs in Edge Functions. See the [Limitations](https://vercel.com/docs/concepts/functions/edge-functions/limitations) section for more information." May be a potential issue.

**Comment by balazsorban44** [comment, marked as answer]
Hi, NextAuth.js (`next-auth`) is not runtime agnostic, but the next generation `@auth/nextjs` will be, which is based on the new `@auth/core`, announced here: https://twitter.com/balazsorban44/status/1603082914362986496

PR here: https://github.com/nextauthjs/next-auth/pull/7443

**Reply by rupampoddar** [reply]
Cool, thanks for the clarification.

**Reply by kzlar** [reply]
Hi @balazsorban44 - appreciate the effort on this library! The `@auth/core` announcement is amazing news. I wanted to ask if this is something that I could try with nextjs right now?

**Reply by dantrain** [reply]
Hi @balazsorban44 do you have a rough timeline for the release of the runtime agnostic `@auth/nextjs` package? Not in a rush, just curious. Really appreciate your work on a great library, thank you!

**Reply by rayli09** [reply]
+1

**Reply by geekyayush** [reply]
Any idea when we will use the new auth/nextjs?
or is there something I can do myself to implement the same?
I'd really appreciate anyone's help, please.

**Reply by re-thc** [reply]
Follow https://github.com/nextauthjs/next-auth/pull/7443 maybe

**Reply by osdiab** [reply]
if you're brave you can use it by installing the `@experimental` tag and the docs linked from that PR.

**Comment by niwatoro** [comment]
I solved the issue using the experimental version as @balazsorban44 suggested. I used Next.js@13.4.6 and NextAuth@0.0.0-manual.e65faa1c.

What I did:

1. Install the experimental version
`npm i next-auth@experimental`

2. Rewrite the code in `app/api/auth/[...nextauth]/route.ts`
```ts
const authOptions: NextAuthConfig = {
  providers: [
    Google({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }) as Provider,
  ],
  secret: process.env.GOOGLE_CLIENT_SECRET,
  pages: {
    signIn: "/signin",
  },
};

const handler = NextAuth(authOptions);
export const GET = handler.handlers.GET;
export const POST = handler.handlers.POST;

export const runtime = "edge";
```

4. See it work!

**Reply by mundume** [reply]
Not working on my codebase though

**Reply by addlistener** [reply]
I got 
```

../../node_modules/next-auth/index.js
Attempted import error: 'raw' is not exported from '@auth/core' (imported as 'raw').

Import trace for requested module:
../../node_modules/next-auth/index.js

../../node_modules/next-auth/index.js
Attempted import error: 'raw' is not exported from '@auth/core' (imported as 'raw').

Import trace for requested module:
../../node_modules/next-auth/index.js
```

=> Found "next-auth@0.0.0-manual.ffd05533"
=> Found "next-auth#@auth/core@0.0.0-manual.59d8da33"

**Reply by weleo** [reply]
same

**Reply by viniciuspalma** [reply]
I got something totally different `Package path . is not exported from package`, someone that had the chance to have this working? I'm using the version `next-auth@0.0.0-manual.ffd05533`

**Reply by JanThiel** [reply]
`Attempted import error: 'raw' is not exported from '@auth/core' (imported as 'raw').`

Try this:
https://github.com/nextauthjs/next-auth/pull/7443#issuecomment-1679731891

**Reply by avdain** [reply]
`import {getServerSession} from "next-auth";` doesn't seem to work with the experimental next-auth, does anyone else have the same problem and does anyone know a workaround?

Mikail

**Reply by JanThiel** [reply]
Check out the docs for v5:

https://auth-docs-git-feat-nextjs-auth-authjs.vercel.app/guides/upgrade-to-v5?authentication-method=api-route#authentication-methods

**Reply by weleo** [reply]
Upgraded my demo applications to v5 without any issues by following the migration guide.

**Comment by avdain** [comment]
Any updates for version 5 @balazsorban44 ?

`import {getServerSession} from "next-auth";` doesn't seem to work with the experimental next-auth version , does anyone else have the same problem and does anyone know a workaround?

**Reply by syhner** [reply]
`lib/auth.ts`
```ts
import NextAuth from 'next-auth';

export const { auth } = NextAuth({ providers: [ /* ... */ ] });
```

Session can be accessed with
```ts
const session = await auth();
```

Also mentioned here: https://github.com/nextauthjs/next-auth/discussions/5855#discussioncomment-6833245 from the reply to your previous message

**Reply by avdain** [reply]
Thanks @syhner, I have already solved the problem by looking at the docs.

**Reply by Bensigo** [reply]
currently getting this error
`` The Edge Function "pages/api/auth/[...nextauth]" must export a `default` function``

my current code 
```js
export {  GET, POST } from "@/server/auth"

export const runtime = "edge";
```

**Reply by colis** [reply]
> currently getting this error `` The Edge Function "pages/api/auth/[...nextauth]" must export a `default` function ``
> 
> my current code
> 
> ```js
> export {  GET, POST } from "@/server/auth"
> 
> export const runtime = "edge";
> ```

Those are route handlers, which are only available inside the `app` directory.

I tried this

```js
import {handlers} from 'auth'

export const config = {
  runtime: 'edge'
}

export default handlers.GET
```
and it works, but you obviously can't export POST from the same module.

Does anyone have a working example of this?

**Reply by jordanvector** [reply]
@colis same issue here, I also have not found a solution

**Comment by azr-arch** [comment]
is my error 
```
Failed to compile.

./node_modules/@babel/runtime/regenerator/index.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime 
Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation

Import trace for requested module:
./node_modules/@babel/runtime/regenerator/index.js
./node_modules/next-auth/react/index.js


> Build failed because of webpack errors
   Creating an optimized production build 
```

same as this thread's ?

**Reply by webstorms** [reply]
you find any solution to this?

**Comment by yangcheng** [comment]
I came to this issue from google search, and after more research I found the new auth.js support edge runtime. https://authjs.dev/guides/edge-compatibility  , migrating to auth.js is working for me.
